### PR TITLE
feat: wire up XVS Bridge

### PR DESCRIPTION
**Pull request by gleiser-oliveira**
## Jira ticket(s)

VEN-2248

## Changes

- Added @venusprotocol/token-bridge dependency
- Added config for the `XVSProxyOFTSrc` and `XVSProxyOFTDest` contracts
- Updated `Bridge` component to use the contracts

### CODE REVIEW

File: src/clients/api/mutations/bridgeXvs/index.ts
Comment (therealemjy) [inline src/clients/api/mutations/bridgeXvs/index.ts]:
```suggestion
  destinationChainId: ChainId;
```

File: src/clients/api/mutations/bridgeXvs/index.ts
Comment (therealemjy) [inline src/clients/api/mutations/bridgeXvs/index.ts]: I feel this name could be confusing: I first thought this corresponded to the gas fee of the transaction made on the chain of origin (which we let users select in their wallet directly).
Perhaps something like `bridgeFeeMantissa` would be clearer. What do you think?

File: src/clients/api/queries/getXvsBridgeFeeEstimation/index.spec.ts
Comment (therealemjy) [inline src/clients/api/queries/getXvsBridgeFeeEstimation/index.spec.ts]: I know most tests currently include the path to the file in their name, but I'd actually advise moving away from this as it is unnecessary. It was originally added to make sure the names of the tests were unique, but that's not necessary as vitest groups them by file anyway.
```suggestion
describe('getXvsBridgeFeeEstimation', () => {
```

File: src/constants/layerZero.ts
Comment (therealemjy) [inline src/constants/layerZero.ts]:
```suggestion
export const LAYER_ZERO_CHAIN_IDS: Record<ChainId, number> = {
```

File: src/constants/layerZero.ts
Comment (therealemjy) [inline src/constants/layerZero.ts]:
```suggestion
export const ADAPTER_PARAMS =
```

File: src/packages/tokens/infos/commonTokens/sepolia.ts
Comment (therealemjy) [inline src/packages/tokens/infos/commonTokens/sepolia.ts]: Do we have duplicates because we're using some sort of temporary XVS contract?

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx]: To prevent ever calling the bridge with an empty address: 
```suggestion
  const onSubmit = () => {
    if (accountAddress) {
      bridgeXvs({
        accountAddress,
        amountMantissa,
        destinationChain: toChainId,
        gasValueMantissa: bridgeEstimatedFeeMantissa,
      });
    }
  };
```

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx:140]: I wouldn't change the names of the properties, as it can reduce reusability (when passing arguments to the function for example).
To prevent conflict with higher-level variables/props, you can instead use the object input directly:
```suggestion
    (input: { fromChainId?: ChainId; toChainId?: ChainId }) => {
```

File: src/pages/Bridge/__tests__/index.spec.tsx
Comment (therealemjy) [inline src/pages/Bridge/__tests__/index.spec.tsx:179]: We'll need to add more tests (for example to check that the mutation is called properly when submitting the form)

File: src/clients/api/mutations/bridgeXvs/useBridgeXvs.ts
Comment (therealemjy) [inline src/clients/api/mutations/bridgeXvs/useBridgeXvs.ts]: Looking at the logic in the Bridge UI, it seems the idea is that on Ethereum and Sepolia we use the dest contract, while on other chains we use the src contract right?

To prevent confusion and also potential bugs, I think we should be more explicit about this logic:

```suggestion
  const tokenBridgeContract = chainId === ChainId.SEPOLIA || chainId === ChainId.ETHEREUM ? tokenBridgeContractDest : tokenBridgeContractSrc;
```

File: src/clients/api/queries/getXvsBridgeFeeEstimation/useGetXvsBridgeFeeEstimation.ts
Comment (therealemjy) [inline src/clients/api/queries/getXvsBridgeFeeEstimation/useGetXvsBridgeFeeEstimation.ts]: Same as previous comment:
```suggestion
  const tokenBridgeContract = chainId === ChainId.SEPOLIA || chainId === ChainId.ETHEREUM ? 
tokenBridgeContractDest : tokenBridgeContractSrc;
```

File: src/clients/api/mutations/bridgeXvs/index.ts
Comment (gleiser-oliveira) [reply src/clients/api/mutations/bridgeXvs/index.ts]: Changed it to `nativeCurrencyFeeMantissa`,  to indicate that the fee will be in BNB, ETH, and so on

File: src/pages/Bridge/index.tsx
Comment (gleiser-oliveira) [reply src/pages/Bridge/index.tsx:140]: I actually kept this, not so much because of the name conflict, but because of the callbacks:

```
callback: () => {
  setValue('fromChainId', newFromChainId); -> changing this to input.fromChainId will throw
},
```
Type narrowing does not help inside the callbacks, so you end up having to add even more code to make sure it knows the key inside `input` has a value. Destructuring `input` would work as expected, but then you have clashing names.

Comment (therealemjy) [reply src/pages/Bridge/index.tsx:140]: I see, that makes sense. Having to assert types would reduce code readability.

File: src/packages/tokens/infos/commonTokens/sepolia.ts
Comment (gleiser-oliveira) [reply src/packages/tokens/infos/commonTokens/sepolia.ts]: Not really, I made this as a temporary fix to Sepolia reporting two XVS addresses as reward tokens, 0x66ebd019E86e0af5f228a0439EBB33f045CBe63E and 0xDb633C11D3F9E6B8D17aC2c972C9e3B05DA59bF9. It will get cleaned up by the contract team, so I removed the obsolete addres (0xDb633C11D3F9E6B8D17aC2c972C9e3B05DA59bF9)

File: src/types/index.ts
Comment (therealemjy) [inline src/types/index.ts]: To be consistent with other global types, I'd lowercase the first letter of each type:
```suggestion
export type TransactionType = 'chain' | 'layerZero';
```

File: src/pages/Bridge/__tests__/index.spec.tsx
Comment (therealemjy) [inline src/pages/Bridge/__tests__/index.spec.tsx]:
```suggestion
  it('lets the user bridge XVS', async () => {
```

File: src/packages/translations/translations/en.json
Comment (therealemjy) [inline src/packages/translations/translations/en.json]: I know you used the copy that's in Figma, but I think we should stay consistent with another similar use case we have in the dApp: the supply/repay forms.
So I'd go with the message:
```suggestion
      "doesNotHaveEnoughNativeFunds": "Insufficient {{tokenSymbol}} balance"
```

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx]: The form schema should not contain fields for errors. Instead, you should handle errors and their messages through the validation rules of each field (in this case with the validation rules of `amountTokens` field): https://zod.dev/?id=strings

So for each of these errors you'd define a rule that you attach to `amountTokens` with a corresponding error message when it is not being respected.

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx]: Following my previous comment about using Zod's error handling, then you'd be able to get rid of this logic (or rather rewrite it into the validation schema) and get the error to display from the form state directly.

File: src/utilities/generateChainExplorerUrl.ts
Comment (therealemjy) [inline src/utilities/generateChainExplorerUrl.ts]: Just to be more explicit: 
```suggestion
export type UrlType = 'address' | 'token' | 'tx' | 'layerZeroTx';
```

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx]: Following my previous comment: this logic should be written into the validation schema of the form.

File: src/constants/chainMetadata.ts
Comment (therealemjy) [inline src/constants/chainMetadata.ts]: To expend the potential use cases for this new field, I'd store the native token directly.
To do that, I'd recommend adding a new `getToken` utility function to the `tokens` package, which could look like this:

```typescript
import { ChainId } from 'types';

import { tokens } from '../../infos/commonTokens';

export interface GetTokenInput {
  chainId: ChainId;
  symbol: string;
}

export const getToken = ({ chainId, symbol }: GetTokenInput) => {
  const chainTokens = tokens[chainId];
  return chainTokens.find(token => token.symbol === symbol);
};
```

You could then use it in the chain metadata:
```suggestion
    nativeToken: getToken({ chainId: ChainId.BSC_MAINNET, symbol: 'BNB' }),
```

With this function built, you could also update the `useGetToken` hook so it uses it in order to reduce repetition.

File: src/constants/usdPriceToken.ts
Comment (therealemjy) [inline src/constants/usdPriceToken.ts]: This feels risky: someone in the future could use it as a reference for a "USD" token.
I understand this is to be able to use `convertPriceMantissaToDollars` in order to convert USD prices expressed with 18 decimals. I'd recommend updating `convertPriceMantissaToDollars` instead, so that instead of taking a `token` field it takes a `decimals` field. This function doesn't need to receive the full token definition since we only use the decimals for the calculation.

**Review by therealemjy — COMMENTED** [review]
When updating the input value, the "Approve token" step appears while the bridge's quote is being fetched, making the UI jump a bit:

https://github.com/VenusProtocol/venus-protocol-interface/assets/44675210/9edfd1a3-2c49-4cfc-8443-809119522bd7

Also, on Sepolia it tells me I don't have enough ETH to send this transaction but the ETH balance in wallet should be sufficient to cover the fee:
Screen Shot 2024-01-15 at 18 33 40
Screen Shot 2024-01-15 at 18 34 57

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx]:
```suggestion
  const [bridgeEstimatedFeeMantissa, bridgeEstimatedFeeTokens] = useMemo(() => {
```

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx:204]: I think we should reset the form when the transaction is successfully sent, to be consistent with the other forms' behavior.

**Review by therealemjy — COMMENTED** [review]
The form is now requesting for XVS approval and disables the input before any amount is entered. To be consistent with other forms, the approve step should only be shown when an amount has been entered in the input.

Screen Shot 2024-01-17 at 09 44 19

File: src/pages/Bridge/useBridgeForm.ts
Comment (therealemjy) [inline src/pages/Bridge/useBridgeForm.ts]:
```suggestion
      amountTokens: '',
```

You need to set this valut to an empty string, otherwise you'll get the following error.
Screen Shot 2024-01-18 at 12 34 24

**Review by therealemjy — APPROVED** [review]
.

**Review by therealemjy — COMMENTED** [review]
.

**Review by therealemjy — COMMENTED** [review]
.

**Review by therealemjy — CHANGES_REQUESTED** [review]
The issue with the approval being asked for at start is gone, but now the forms remains disabled after giving approval:
Screen Shot 2024-01-18 at 12 35 49

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx]:
```suggestion
  const [fromChainIdOptions, toChainIdOptions] = useMemo(() => {
```

A small tip, this VSCode extension can check spelling errors in your code: https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

File: src/pages/Bridge/index.tsx
Comment (therealemjy) [inline src/pages/Bridge/index.tsx:271]: Unfortunately this logic is causing some trouble. If you connect your wallet to Sepolia, go to the Bridge page then change your wallet network to BSC testnet you end up with the following UI:
Screen Shot 2024-01-19 at 10 10 33

Moreover, if you do not list all the available chains, then users won't be able to perform all the actions they should be able to.
Here's a screenshot that illustrates what I mean:
Screen Shot 2024-01-19 at 09 59 19

In that situation, my wallet is connected to the opBNB testnet network when I open the bridge. With this, I cannot bridge XVS from opBNB testnet to Sepolia, or from BNB testnet to opBNB testnet without changing the chain from the header or my wallet directly.

I recommend keeping the original UX:
- all chains are listed in both selects (the logic inside `handleChainFieldChange` will take care of conflicts when changing chains)
- when opening the bridge, the "from" chain corresponds to the current chain of the dApp and the "to" chain corresponds to the next available chain in the list

Comment (gleiser-oliveira) [reply src/pages/Bridge/index.tsx:271]: Do let me know if this works for you, here's how I tackled this issue:
- I'm still filtering the options based on which chain is the selected source or destiny. In other words, it's not possible to select Sepolia as "from" and opBNB as "to"
- If the chain is changed directly in the header dropdown, the form reacts to it by making it be the "from" chain.

**Review by therealemjy — APPROVED** [review]

**Pull request merged**
